In [1]:
import pandas as pd

In [2]:
# read the dataset
df = pd.read_csv('../data/RTAQI.csv')
df.sample(5)

country           state       city                             station  \
1334   India           Bihar    Bettiah    Kamalnath Nagar, Bettiah - BSPCB   
2337   India          Odisha      Angul           Hakimapada, Angul - OSPCB   
2577   India  Andhra_Pradesh  Amaravati      Secretariat, Amaravati - APPCB   
2275   India     Maharashtra     Mumbai    Navy Nagar-Colaba, Mumbai - IITM   
1663   India     Maharashtra     Nagpur  Opp GPO Civil Lines, Nagpur - MPCB   

              last_update   latitude  longitude pollutant_id  pollutant_min  \
1334  14-04-2024 11:00:00  26.803650  84.519540          SO2            NaN   
2337  14-04-2024 11:00:00  20.832874  85.104082          NO2            1.0   
2577  14-04-2024 11:00:00  16.515083  80.518167           CO           24.0   
2275  14-04-2024 11:00:00  18.897756  72.813320          NO2            NaN   
1663  14-04-2024 11:00:00  21.152875  79.051753          SO2            2.0   

      pollutant_max  pollutant_avg  
1334            NaN            NaN  
2337            1.0            1.0  
2577          114.0           47.0  
2275            NaN            NaN  
1663           17.0            7.0

In [3]:
# check dataframe information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3233 entries, 0 to 3232
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   country        3233 non-null   object 
 1   state          3233 non-null   object 
 2   city           3233 non-null   object 
 3   station        3233 non-null   object 
 4   last_update    3233 non-null   object 
 5   latitude       3233 non-null   float64
 6   longitude      3233 non-null   float64
 7   pollutant_id   3233 non-null   object 
 8   pollutant_min  3030 non-null   float64
 9   pollutant_max  3030 non-null   float64
 10  pollutant_avg  3030 non-null   float64
dtypes: float64(5), object(6)
memory usage: 278.0+ KB


There appears to be missing values in column `pollutant_min`, `pollutant_max`, `pollutant_avg`

In [4]:
# checking dataframe statistics
df.describe()

latitude    longitude  pollutant_min  pollutant_max  pollutant_avg
count  3233.000000  3233.000000    3030.000000    3030.000000    3030.000000
mean     22.954409    78.541401      23.398680      81.019472      41.901650
std       5.325533     4.863931      25.261122      92.897985      43.427336
min       8.514909    70.909168       1.000000       1.000000       1.000000
25%      19.110740    75.565602       6.000000      19.000000      11.000000
50%      23.717634    77.287209      15.000000      50.000000      29.000000
75%      27.308328    80.327067      33.000000     103.000000      56.000000
max      31.620000    94.636574     306.000000     500.000000     340.000000

**Pollutant Min:**
The mean minimum pollutant value is approximately 23.40, indicating the average lowest pollutant concentration across observations. <br>
The minimum pollutant minimum value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant minimum value of 306 indicates relatively high minimum pollutant levels in some observations. <br>
The standard deviation of approximately 25.26 reflects the variability in minimum pollutant values across the dataset, showcasing fluctuating pollution levels. <br>

**Pollutant Max:**
The mean maximum pollutant value is approximately 81.02, indicating the average highest pollutant concentration across observations. <br>
The minimum pollutant maximum value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant maximum value of 500 indicates relatively high maximum pollutant levels in some observations. <br>
The standard deviation of approximately 92.90 reflects the variability in maximum pollutant values across the dataset, highlighting diverse pollution levels. <br>

**Pollutant Avg:**
The mean average pollutant value is approximately 41.90, indicating the average pollutant concentration across observations. <br>
The minimum pollutant average value of 1 suggests the presence of outliers or very low pollutant levels in some observations. <br>
The maximum pollutant average value of 340 indicates relatively high average pollutant levels in some observations. <br>
The standard deviation of approximately 43.43 reflects the variability in average pollutant values across the dataset, illustrating fluctuations in pollution levels over time or across locations. <br>

In [5]:
# checking for uniques in country
df.country.unique()

# since there is only one country in this dataset, we can drop this field
df.shape
df = df.drop('country', axis=1)
df.shape

(3233, 10)

In [6]:
# convert last_update to datetime
df['last_update'] = pd.to_datetime(df['last_update'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3233 entries, 0 to 3232
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   state          3233 non-null   object        
 1   city           3233 non-null   object        
 2   station        3233 non-null   object        
 3   last_update    3233 non-null   datetime64[ns]
 4   latitude       3233 non-null   float64       
 5   longitude      3233 non-null   float64       
 6   pollutant_id   3233 non-null   object        
 7   pollutant_min  3030 non-null   float64       
 8   pollutant_max  3030 non-null   float64       
 9   pollutant_avg  3030 non-null   float64       
dtypes: datetime64[ns](1), float64(5), object(4)
memory usage: 252.7+ KB


C:\Users\Asif Sayyed\AppData\Local\Temp\ipykernel_20092\1416567406.py:2: UserWarning: Parsing dates in %d-%m-%Y %H:%M:%S format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df['last_update'] = pd.to_datetime(df['last_update'])


In [7]:
# glancing over the dataset
df.sample(5)

state         city                                 station  \
432      Odisha      Cuttack               CDA Area, Cuttack - OSPCB   
532   TamilNadu      Chennai             Arumbakkam, Chennai - TNPCB   
2388  Rajasthan  Chittorgarh      Shastri Nagar, Chittorgarh - RSPCB   
2721    Gujarat        Vatva              Phase-4 GIDC, Vatva - GPCB   
1420      Delhi        Delhi  Jawaharlal Nehru Stadium, Delhi - DPCC   

             last_update   latitude  longitude pollutant_id  pollutant_min  \
432  2024-04-14 11:00:00  20.488910  85.847680          SO2            1.0   
532  2024-04-14 11:00:00  13.066400  80.211200          SO2            1.0   
2388 2024-04-14 11:00:00  24.892047  74.623527         PM10           65.0   
2721 2024-04-14 11:00:00  22.969611  72.643500        PM2.5           44.0   
1420 2024-04-14 11:00:00  28.580280  77.233829           CO           34.0   

      pollutant_max  pollutant_avg  
432            48.0           18.0  
532             6.0            4.0  
2388          142.0           97.0  
2721           63.0           54.0  
1420           69.0           41.0

In [8]:
# checking for missing values
df.isnull().sum()

state              0
city               0
station            0
last_update        0
latitude           0
longitude          0
pollutant_id       0
pollutant_min    203
pollutant_max    203
pollutant_avg    203
dtype: int64

We have 203 values missing for all three pollutant numerical values.

In [9]:
# checking for any pattern in missing values
df.sample(10)

state       city                                         station  \
1391  Chhattisgarh   Milupara  Govt. Higher Secondary School, Milupara - CECB   
2415     Rajasthan  Jhunjhunu                  Indra Nagar, Jhunjhunu - RSPCB   
1978         Bihar    Katihar                    Mirchaibari, Katihar - BSPCB   
126          Delhi      Delhi                       Mandir Marg, Delhi - DPCC   
3226   West_Bengal    Kolkata               Rabindra Sarobar, Kolkata - WBPCB   
1440         Delhi      Delhi                               Pusa, Delhi - IMD   
2398     Rajasthan     Jaipur          Police Commissionerate, Jaipur - RSPCB   
3091     TamilNadu       Ooty                     Bombay Castel, Ooty - TNPCB   
118          Delhi      Delhi                               ITO, Delhi - CPCB   
706          Bihar      Patna      Govt. High School Shikarpur, Patna - BSPCB   

             last_update   latitude  longitude pollutant_id  pollutant_min  \
1391 2024-04-14 11:00:00  22.191017  83.519701          NH3            NaN   
2415 2024-04-14 11:00:00  28.108988  75.386577        PM2.5           33.0   
1978 2024-04-14 11:00:00  25.560083  87.553265         PM10           68.0   
126  2024-04-14 11:00:00  28.636429  77.201067         PM10           50.0   
3226 2024-04-14 11:00:00  22.511060  88.351420         PM10           72.0   
1440 2024-04-14 11:00:00  28.636110  77.173332           CO           75.0   
2398 2024-04-14 11:00:00  26.916409  75.799490          SO2           13.0   
3091 2024-04-14 11:00:00  11.406829  76.713897         PM10            NaN   
118  2024-04-14 11:00:00  28.628624  77.241060        PM2.5           64.0   
706  2024-04-14 11:00:00  25.592539  85.227158          SO2            5.0   

      pollutant_max  pollutant_avg  
1391            NaN            NaN  
2415          187.0           83.0  
1978          215.0          125.0  
126           250.0           93.0  
3226          104.0           89.0  
1440           94.0           88.0  
2398           28.0           21.0  
3091            NaN            NaN  
118           302.0           95.0  
706            11.0            9.0

When we have missing values, we have it for the entire row of data, likey due to sensor malfunctions, we can impute this data with mean to deal with them.

In [10]:
# checking for all the unique pollutants in the dataset
df.pollutant_id.unique()

array(['NH3', 'SO2', 'PM2.5', 'PM10', 'OZONE', 'CO', 'NO2'], dtype=object)

In [11]:
# List of all possible pollutants
possible_pollutants = ['NH3', 'SO2', 'PM2.5', 'PM10', 'OZONE', 'CO', 'NO2']

def impute_missing_values(df, possible_pollutants):
    # Group by 'pollutant_id' and calculate the mean pollutant values for each group
    pollutant_means = df.groupby('pollutant_id')[['pollutant_min', 'pollutant_avg', 'pollutant_max']].mean().to_dict(orient='index')
    
    # Function to impute missing values based on pollutant category means
    def impute_missing(row):
        pollutant_id = row['pollutant_id']
        if pd.isnull(row['pollutant_min']):
            row['pollutant_min'] = pollutant_means.get(pollutant_id, {}).get('pollutant_min', None)
        if pd.isnull(row['pollutant_avg']):
            row['pollutant_avg'] = pollutant_means.get(pollutant_id, {}).get('pollutant_avg', None)
        if pd.isnull(row['pollutant_max']):
            row['pollutant_max'] = pollutant_means.get(pollutant_id, {}).get('pollutant_max', None)
        return row
    
    # Apply the imputation function to each row of the DataFrame
    df_imputed = df.apply(impute_missing, axis=1)
    
    return df_imputed

In [12]:
df_imputed = impute_missing_values(df, possible_pollutants)

In [13]:
df.describe()

last_update     latitude    longitude  pollutant_min  \
count                           3233  3233.000000  3233.000000    3030.000000   
mean   2024-04-14 10:59:59.999999744    22.954409    78.541401      23.398680   
min              2024-04-14 11:00:00     8.514909    70.909168       1.000000   
25%              2024-04-14 11:00:00    19.110740    75.565602       6.000000   
50%              2024-04-14 11:00:00    23.717634    77.287209      15.000000   
75%              2024-04-14 11:00:00    27.308328    80.327067      33.000000   
max              2024-04-14 11:00:00    31.620000    94.636574     306.000000   
std                              NaN     5.325533     4.863931      25.261122   

       pollutant_max  pollutant_avg  
count    3030.000000    3030.000000  
mean       81.019472      41.901650  
min         1.000000       1.000000  
25%        19.000000      11.000000  
50%        50.000000      29.000000  
75%       103.000000      56.000000  
max       500.000000     340.000000  
std        92.897985      43.427336

In [14]:
df_imputed.describe()

last_update     latitude    longitude  pollutant_min  \
count                           3233  3233.000000  3233.000000    3233.000000   
mean   2024-04-14 10:59:59.999999744    22.954409    78.541401      23.538871   
min              2024-04-14 11:00:00     8.514909    70.909168       1.000000   
25%              2024-04-14 11:00:00    19.110740    75.565602       6.000000   
50%              2024-04-14 11:00:00    23.717634    77.287209      15.965909   
75%              2024-04-14 11:00:00    27.308328    80.327067      34.000000   
max              2024-04-14 11:00:00    31.620000    94.636574     306.000000   
std                              NaN     5.325533     4.863931      24.871403   

       pollutant_max  pollutant_avg  
count    3233.000000    3233.000000  
mean       81.495185      42.171106  
min         1.000000       1.000000  
25%        20.000000      12.000000  
50%        51.000000      29.000000  
75%       105.000000      58.000000  
max       500.000000     340.000000  
std        91.543673      42.921555

We have imputed all the missing values without making a significant difference in the central tendencies

In [15]:
# exporting the data
df_imputed.to_csv("../data/RTAQI_update.csv", index=False)